// mindee/src/product/us/w9/w9V1Page.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.W9V1Page = void 0;
const common_1 = require("../../../parsing/common");
const standard_1 = require("../../../parsing/standard");
const w9V1Document_1 = require("./w9V1Document");
/**
 * Page data for US W9, API version 1.
 */
class W9V1Page extends w9V1Document_1.W9V1Document {
    constructor(rawPrediction, pageId) {
        super();
        this.address = new standard_1.StringField({
            prediction: rawPrediction["address"],
            pageId: pageId,
        });
        this.businessName = new standard_1.StringField({
            prediction: rawPrediction["business_name"],
            pageId: pageId,
        });
        this.cityStateZip = new standard_1.StringField({
            prediction: rawPrediction["city_state_zip"],
            pageId: pageId,
        });
        this.ein = new standard_1.StringField({
            prediction: rawPrediction["ein"],
            pageId: pageId,
        });
        this.name = new standard_1.StringField({
            prediction: rawPrediction["name"],
            pageId: pageId,
        });
        this.signatureDatePosition = new standard_1.PositionField({
            prediction: rawPrediction["signature_date_position"],
            pageId: pageId,
        });
        this.signaturePosition = new standard_1.PositionField({
            prediction: rawPrediction["signature_position"],
            pageId: pageId,
        });
        this.ssn = new standard_1.StringField({
            prediction: rawPrediction["ssn"],
            pageId: pageId,
        });
        this.taxClassification = new standard_1.StringField({
            prediction: rawPrediction["tax_classification"],
            pageId: pageId,
        });
        this.taxClassificationLlc = new standard_1.StringField({
            prediction: rawPrediction["tax_classification_llc"],
            pageId: pageId,
        });
        this.taxClassificationOtherDetails = new standard_1.StringField({
            prediction: rawPrediction["tax_classification_other_details"],
            pageId: pageId,
        });
        this.w9RevisionDate = new standard_1.StringField({
            prediction: rawPrediction["w9_revision_date"],
            pageId: pageId,
        });
    }
    toString() {
        const outStr = `:Name: ${this.name}
:SSN: ${this.ssn}
:Address: ${this.address}
:City State Zip: ${this.cityStateZip}
:Business Name: ${this.businessName}
:EIN: ${this.ein}
:Tax Classification: ${this.taxClassification}
:Tax Classification Other Details: ${this.taxClassificationOtherDetails}
:W9 Revision Date: ${this.w9RevisionDate}
:Signature Position: ${this.signaturePosition}
:Signature Date Position: ${this.signatureDatePosition}
:Tax Classification LLC: ${this.taxClassificationLlc}`.trimEnd();
        return (0, common_1.cleanOutString)(outStr);
    }
}
exports.W9V1Page = W9V1Page;
