// mindee/src/product/internal.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.us = exports.receipt = exports.proofOfAddress = exports.passport = exports.multiReceiptsDetector = exports.invoiceSplitter = exports.invoice = exports.fr = exports.financialDocument = exports.eu = exports.custom = exports.cropper = exports.barcodeReader = void 0;
exports.barcodeReader = __importStar(require("./barcodeReader/internal"));
exports.cropper = __importStar(require("./cropper/internal"));
exports.custom = __importStar(require("./custom/internal"));
exports.eu = __importStar(require("./eu/internal"));
exports.financialDocument = __importStar(require("./financialDocument/internal"));
exports.fr = __importStar(require("./fr/internal"));
exports.invoice = __importStar(require("./invoice/internal"));
exports.invoiceSplitter = __importStar(require("./invoiceSplitter/internal"));
exports.multiReceiptsDetector = __importStar(require("./multiReceiptsDetector/internal"));
exports.passport = __importStar(require("./passport/internal"));
exports.proofOfAddress = __importStar(require("./proofOfAddress/internal"));
exports.receipt = __importStar(require("./receipt/internal"));
exports.us = __importStar(require("./us/internal"));
